snomed/com.b2international.snowowl.snomed.datastore/src/com/b2international/snowowl/snomed/datastore/index/change/ComponentEffectiveTimeRestoreChangeProcessor.java

```diff
@@ -201,85 +190,29 @@ private boolean canRestoreEffectiveTime(SnomedDocument componentToRestore, Snome
-		final CodeSystem workingCodeSystem = relativeCodeSystems.stream().findFirst().get();
-
-		CodeSystemVersionSearchRequestBuilder versionSearch = CodeSystemRequests.prepareSearchCodeSystemVersion()
-				.one()
-				.filterByCodeSystemShortName(workingCodeSystem.getShortName())
-				.sortBy(SearchResourceRequest.SortField.descending(CodeSystemVersionEntry.Fields.EFFECTIVE_DATE));
+	private List<String> getAvailableVersionPaths(RepositoryContext context, String branchPath) {
+		final List<CodeSystemURI> codeSystemsToCheck = Lists.newArrayList();
 		
-		// if specified and not restoring effective time on the Code System Working Branch then filter by created at up until the specified branch base timestamp
-		if (branchBaseTimestamp > 0L && !branchPath.equals(workingCodeSystem.getBranchPath())) {
-			versionSearch.filterByCreatedAt(0L, branchBaseTimestamp);
-		}
+		CodeSystem relativeCodeSystem = context.service(RepositoryCodeSystemProvider.class).get(branchPath);
 		
-		final Optional<CodeSystemVersionEntry> workingCodeSystemVersion = versionSearch
-				.build(SnomedDatastoreActivator.REPOSITORY_UUID)
-				.execute(bus)
-				.getSync()
-				.first();
-
-		final List<CodeSystemVersionEntry> relativeCodeSystemVersions = Lists.newArrayList();
-
-		if (workingCodeSystemVersion.isPresent() && !Strings.isNullOrEmpty(workingCodeSystemVersion.get().getPath())) {
-			relativeCodeSystemVersions.add(workingCodeSystemVersion.get());
+		// in case of an upgrade CodeSystem check the original CodeSystem as well
+		if (relativeCodeSystem.getUpgradeOf() != null) {
+			codeSystemsToCheck.add(CodeSystemURI.latest(relativeCodeSystem.getUpgradeOf().getCodeSystem()));
+		} else {
+			// in case of regular CodeSystem check that
+			codeSystemsToCheck.add(CodeSystemURI.latest(relativeCodeSystem.getCodeSystemURI().getCodeSystem()));
 		}
-
-		if (relativeCodeSystems.size() > 1) {
-
-			relativeCodeSystems.stream().skip(1).forEach(codeSystem -> {
-
-				final Map<String, CodeSystemVersionEntry> pathToVersionMap = CodeSystemRequests.prepareSearchCodeSystemVersion()
-						.all()
-						.filterByCodeSystemShortName(codeSystem.getShortName())
-						.build(SnomedDatastoreActivator.REPOSITORY_UUID)
-						.execute(bus)
-						.getSync(1, TimeUnit.MINUTES)
-						.stream()
-						.collect(Collectors.toMap(version -> version.getPath(), v -> v));
-
-				final Iterator<IBranchPath> branchPathIterator = BranchPathUtils.bottomToTopIterator(BranchPathUtils.createPath(branchPath));
-
-				while (branchPathIterator.hasNext()) {
-					final IBranchPath candidate = branchPathIterator.next();
-					if (pathToVersionMap.containsKey(candidate.getPath())) {
-						relativeCodeSystemVersions.add(pathToVersionMap.get(candidate.getPath()));
-						break;
-					}
-				}
-
-			});
-
+		
+		// always check the direct extensionOf (aka parent) CodeSystem
+		if (relativeCodeSystem.getExtensionOf() != null) {
+			if (relativeCodeSystem.getExtensionOf().isHead()) {
```

Comment:
In `CodeSystemCreateRequest`, `extensionOf` is always mapped to a code system version entry of the parent code system, this should not happen:

https://github.com/b2ihealthcare/snow-owl/blob/c8067e26c1610c2c6ad86f6994b90cffe7d8c93e/core/com.b2international.snowowl.core/src/com/b2international/snowowl/core/codesystem/CodeSystemCreateRequest.java#L172-L175

Reply:
Right, we can remove the if block then. 😄

Reply:
Unfortunately, no, this might not be true in older systems, we need to keep this for now.